src/main/java/org/junitpioneer/jupiter/RepeatFailedTestExtension.java

```diff
@@ -0,0 +1,128 @@
+		// the annotation only applies to methods
+		Method testMethod = context.getRequiredTestMethod();
+		return isAnnotated(testMethod, RepeatFailedTest.class);
+	}
+
+	@Override
+	public Stream<TestTemplateInvocationContext> provideTestTemplateInvocationContexts(ExtensionContext context) {
+		FailedTestRepeater repeater = repeaterFor(context);
+		return stream(spliteratorUnknownSize(repeater, ORDERED), false);
+	}
+
+	@Override
+	public void handleTestExecutionException(ExtensionContext context, Throwable throwable) throws Throwable {
+		// this `context` (M) is a child of the context passed to `provideTestTemplateInvocationContexts` (T),
+		// which means M's store content is invisible to T's store; this can be fixed by using T's store here
+		//@formatter:off
+		ExtensionContext templateContext = context
+				.getParent()
+				.orElseThrow(
+						() -> new IllegalStateException(
+								"Extension context \"" + context + "\" should have a parent context."));
+		//@formatter:on
+		repeaterFor(templateContext).failed(throwable);
+	}
+
+	private static FailedTestRepeater repeaterFor(ExtensionContext context) {
+		//@formatter:off
+		Method repeatedTest = context.getRequiredTestMethod();
+		return context
+				.getStore(NAMESPACE)
+				.getOrComputeIfAbsent(
+						repeatedTest.toString(),
+						__ -> FailedTestRepeater.createFor(repeatedTest),
+						FailedTestRepeater.class);
+		//@formatter:on
+	}
+
+	private static class FailedTestRepeater implements Iterator<RepeatFailedTestInvocationContext> {
+
+		private final int maxRepetitions;
+
+		private int repetitionsSoFar;
+		private int exceptionsSoFar;
+
+		private FailedTestRepeater(int maxRepetitions) {
+			this.maxRepetitions = maxRepetitions;
+			this.repetitionsSoFar = 0;
+			this.exceptionsSoFar = 0;
+		}
+
+		static FailedTestRepeater createFor(Method repeatedTest) {
+			//@formatter:off
+			RepeatFailedTest repeatFailedTest = findAnnotation(repeatedTest, RepeatFailedTest.class)
+					.orElseThrow(() -> new IllegalStateException("@RepeatFailedTest is missing."));
+			return new FailedTestRepeater(repeatFailedTest.value());
+			//@formatter:on
+		}
+
+		void failed(Throwable exception) throws Throwable {
+			exceptionsSoFar++;
+
+			boolean allRepetitionsFailed = exceptionsSoFar == maxRepetitions;
+			if (allRepetitionsFailed)
+				throw new AssertionError(
+					format("Test execution #%d (of up to %d) failed ~> test fails", exceptionsSoFar, maxRepetitions));
+			else
+				throw new TestAbortedException(
+					format("Test execution #%d (of up to %d) failed ~> will retry...", exceptionsSoFar, maxRepetitions),
+					exception);
+		}
+
+		@Override
+		public boolean hasNext() {
+			// there's always at least one execution
+			if (repetitionsSoFar == 0)
+				return true;
+
+			// if we caught an exception in each repetition, each repetition failed, including the previous one
+			boolean previousFailed = repetitionsSoFar == exceptionsSoFar;
```

Comment:
In the case of parallel execution this should block until the previous invocation is done. Alternatively, we could meta-annotate `@RepeatFailedTest` with `@Execution(SAME_THREAD)`. However, the latter would cause all `@RepeatFailedTest` methods in a test class to run sequentially.

Reply:
I think this extension thread-safe in a way that makes sense and interacts well with Jupiter is not trivial and should be done in a separate issue. My questions:

* should all repetitions be executed in parallel?
* if not, would a block during the traversal of the stream returned by `provideTestTemplateInvocationContexts` not be unexpected for Jupiter?

(While thinking about this, I wanted to write a test that executes `@RepeatFailedTest` in parallel. Is there a good way to do that?)

For now, I want to go with `@Execution(SAME_THREAD)`.

Reply:
Oh, that requires bumping the Jupiter dependency from 5.1.1 to 5.3. I'm okay with that. Does anybody have any qualms about this?